src/hotspot/share/gc/shared/partialArrayState.cpp

```diff
@@ -81,73 +62,79 @@ struct PartialArrayStateAllocator::Impl::FreeListEntry {
+  : _manager(manager),
+    _free_list(),
+    _arena(manager->register_allocator())
+{}
 
-PartialArrayStateAllocator::Impl::~Impl() {
-  // We don't need to clean up the free lists.  Deallocating the entries
+PartialArrayStateAllocator::~PartialArrayStateAllocator() {
+  // We don't need to clean up the free list.  Deallocating the entries
   // does nothing, since we're using arena allocation.  Instead, leave it
-  // to the arena destructor to release the memory.
-  FREE_C_HEAP_ARRAY(FreeListEntry*, _free_lists);
-  for (uint i = 0; i < _num_workers; ++i) {
-    _arenas[i].~Arena();
-  }
-  FREE_C_HEAP_ARRAY(Arena*, _arenas);
+  // to the manager to release the memory.
+  // Inform the manager that an allocator is no longer in use.
+  _manager->release_allocator();
 }
 
-PartialArrayState* PartialArrayStateAllocator::Impl::allocate(uint worker_id,
-                                                              oop src, oop dst,
-                                                              size_t index,
-                                                              size_t length,
-                                                              size_t initial_refcount) {
+PartialArrayState* PartialArrayStateAllocator::allocate(oop src, oop dst,
+                                                        size_t index,
+                                                        size_t length,
+                                                        size_t initial_refcount) {
   void* p;
-  FreeListEntry* head = _free_lists[worker_id];
+  FreeListEntry* head = _free_list;
   if (head == nullptr) {
-    p = NEW_ARENA_OBJ(&_arenas[worker_id], PartialArrayState);
+    p = NEW_ARENA_OBJ(_arena, PartialArrayState);
   } else {
-    _free_lists[worker_id] = head->_next;
+    _free_list = head->_next;
     head->~FreeListEntry();
     p = head;
   }
   return ::new (p) PartialArrayState(src, dst, index, length, initial_refcount);
 }
 
-void PartialArrayStateAllocator::Impl::release(uint worker_id, PartialArrayState* state) {
+void PartialArrayStateAllocator::release(PartialArrayState* state) {
   size_t refcount = Atomic::sub(&state->_refcount, size_t(1), memory_order_release);
   if (refcount != 0) {
     assert(refcount + 1 != 0, "refcount underflow");
   } else {
-    OrderAccess::acquire();
-    state->~PartialArrayState();
-    _free_lists[worker_id] = ::new (state) FreeListEntry(_free_lists[worker_id]);
+    // Don't need to call destructor; can't if not destructible.
+    static_assert(!std::is_destructible<PartialArrayState>::value, "expected");
+    _free_list = ::new (state) FreeListEntry(_free_list);
   }
 }
 
-PartialArrayStateAllocator::PartialArrayStateAllocator(uint num_workers)
-  : _impl(new Impl(num_workers))
+PartialArrayStateManager::PartialArrayStateManager(uint max_allocators)
+  : _arenas(NEW_C_HEAP_ARRAY(Arena, max_allocators, mtGC)),
+    _max_allocators(max_allocators),
+    _registered_allocators(0),
+    _released_allocators(0)
```

Comment:
`_released_allocators` is a debug only variable, should fail in release build

Reply:
Well spotted.  It seems I haven't done a release build since my final touch-up to make it debug-only.
I'll push an update once it's been through our CI.